contrib/babelfishpg_tsql/src/pltsql_coerce.c

```diff
@@ -885,6 +912,164 @@ run_tsql_best_match_heuristics(int nargs, Oid *input_typeids, FuncCandidateList
 	return new_candidates;
 }
 
+/*
+ * For a given function details, validate whether it is in exception function list
+ * and also validate the input argument data types.
+ */
+bool validate_exception_function(char *func_nsname, char *func_name, int nargs, Oid *input_typeids)
+{
+	tsql_exception_function_t  *exception_func = NULL;
+	int							i;
+	bool						type_match;
+	Oid							inputTypId, validTypId;
+	Oid							sys_varcharoid;
+
+	/* Sanity checks */
+	if (func_name == NULL || (nargs != 0 && input_typeids == NULL))
+		return false;
+
+	/* 
+	 * Exception function handling is only for system functions.
+	 * If func_nsname is NULL, consider it to be a "sys".
+	 */
+	if (func_nsname != NULL &&
+		(strlen(func_nsname) != 3 || strncmp(func_nsname, "sys", 3) != 0))
+		return false;
+
+	/* Get Exception function details */
+	for (i = 0; i < TOTAL_TSQL_EXCEPTION_FUNCTION_COUNT; i++)
+	{
+		exception_func = &tsql_exception_function_list[i];
+		if (strcmp(func_name, exception_func->funcname) == 0
+			&& nargs == exception_func->nargs)
+				break;
+		exception_func = NULL;
+	}
+
+	/* If function is not an exception function no additional handling required */
+	if (i >= TOTAL_TSQL_EXCEPTION_FUNCTION_COUNT
+		|| exception_func == NULL)
+	{
+		return false;
+	}
+
+	sys_varcharoid = get_sys_varcharoid();
+	/* Report error in case of invalid argument datatype */
+	for (i = 0; i < exception_func->nargs; i++)
+	{
+		/* for unknown literals consider its type as sys.VARCHAR */
+		inputTypId = (input_typeids[i] == UNKNOWNOID) ? sys_varcharoid : input_typeids[i];
+
+		type_match = false;
+		for (int j = 0; j < exception_func->valid_arg_types[i].len; j++)
+		{
+			if (!OidIsValid(exception_func->valid_arg_types[i].valid_types_oid[j]))
+				exception_func->valid_arg_types[i].valid_types_oid[j] = 
+					(*common_utility_plugin_ptr->lookup_tsql_datatype_oid) (exception_func->valid_arg_types[i].valid_types[j]);
```

Comment:
what if common_utility_plugin_ptr is not available? result in crash

Reply:
added check to throw error in case common_utility_plugin_ptr is not available.